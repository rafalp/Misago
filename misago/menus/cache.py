from django.core.cache import cache

from ..cache.versions import invalidate_cache
from . import MENU_ITEMS_CACHE


def get_menus_cache(cache_versions):
    key = get_cache_key(cache_versions)
    return cache.get(key)


def set_menus_cache(cache_versions, menus):
    key = get_cache_key(cache_versions)
    cache.set(key, menus)


def get_cache_key(cache_versions):
    return "%s_%s" % (MENU_ITEMS_CACHE, cache_versions[MENU_ITEMS_CACHE])


def clear_menus_cache():
    invalidate_cache(MENU_ITEMS_CACHE)
