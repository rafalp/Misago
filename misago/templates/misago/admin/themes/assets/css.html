{% load i18n %}
<div class="card card-admin-table mb-3">
  <div class="card-body">
    <div class="row align-items-center">
      <div class="col">
        <h4 class="card-title">
          {% trans "CSS files" %}
        </h4>
      </div>
      <div class="col-auto">
        <button type="button" class="btn btn-light btn-sm" data-toggle="modal" data-target="#uploadCss">
          {% trans "Upload" %}
        </button>
      </div>
      <div class="col-auto">
        <a href="{% url 'misago:admin:themes:new-css-link' pk=theme.pk %}" class="btn btn-light btn-sm">
          {% trans "Link" %}
        </a>
      </div>
      <div class="col-auto">
        <a href="{% url 'misago:admin:themes:new-css-file' pk=theme.pk %}" class="btn btn-light btn-sm">
          {% trans "Create" %}
        </a>
      </div>
    </div>
  </div>
  {% with theme.css.all as css %}
    <table class="table">
      <tr>
        <th style="width: 1px;">&nbsp;</th>
        <th style="width: 250px;">{% trans "Name" %}</th>
        <th style="width: 250px;">{% trans "Modified" %}</th>
        <th>{% trans "Size" %}</th>
        <th style="width: 1px;">&nbsp;</th>
        <th style="width: 1px;">&nbsp;</th>
        <th style="width: 1px;">&nbsp;</th>
        <th style="width: 1px;">
          <input type="checkbox" {% if not css %}disabled{% endif %}>
        </th>
      </tr>
      {% for item in css %}
        <tr>
          <td class="pr-0">
            <a class="btn btn-light btn-sm" href="{% if item.url %}{{ item.url }}{% else %}{{ item.source_file.url }}{% endif %}" target="blank">
              <span class="far fa-file"></span>
            </a>
          </td>
          <td class="small">
            {% if item.url %}
              <a href="{% url 'misago:admin:themes:edit-css-link' pk=theme.pk css_pk=item.pk %}" class="item-name">
                {{ item }}
              </a>
            {% else %}
              <a href="{% url 'misago:admin:themes:edit-css-file' pk=theme.pk css_pk=item.pk %}" class="item-name">
                {{ item }}
              </a>
            {% endif %}
          </td>
          <td class="small">
            <abbr data-timestamp="{{ item.modified_on.isoformat }}" data-format="LL">
              {{ item.modified_on }}
            </abbr>
          </td>
          <td class="small">
            {{ item.size|filesizeformat }}
          </td>
          <td>
            <button class="btn btn-light btn-sm btn-move" form="move-down-{{ item.pk }}" data-tooltip="top" title="{% trans 'Move down' %}" {% if forloop.last %}disabled{% endif %}>
              <span class="fa fa-chevron-down"></span>
            </button>
          </td>
          <td>
            <button class="btn btn-light btn-sm btn-move" form="move-up-{{ item.pk }}" data-tooltip="top" title="{% trans 'Move up' %}" {% if forloop.first %}disabled{% endif %}>
              <span class="fa fa-chevron-up"></span>
            </button>
          </td>
          <td>
            {% if item.url %}
              <a href="{% url 'misago:admin:themes:edit-css-link' pk=theme.pk css_pk=item.pk %}" class="btn btn-light btn-sm">
                {% trans "Edit" %}
              </a>
            {% else %}
              <a href="{% url 'misago:admin:themes:edit-css-file' pk=theme.pk css_pk=item.pk %}" class="btn btn-light btn-sm">
                {% trans "Edit" %}
              </a>
            {% endif %}
          </td>
          <td class="row-select">
            <input type="checkbox" form="delete-css" name="item" value="{{ item.pk }}">
          </td>
        </tr>
      {% empty %}
        <tr class="blankslate">
          <td colspan="8">{% trans "This theme has no CSS files." %}</td>
        </tr>
      {% endfor %}
    </table>
    {% if css %}
      <div class="card-body text-right">
        <form id="delete-css" action="{% url 'misago:admin:themes:delete-css' pk=theme.pk %}" method="post">
          {% csrf_token %}
          <button class="btn btn-light btn-sm" disabled>
            {% trans "Delete selected" %}
          </button>
        </form>
      </div>
    {% endif %}
    {% for item in css %}
      <form action="{% url 'misago:admin:themes:move-css-up' pk=theme.pk css_pk=item.pk %}" method="post" id="move-up-{{ item.pk }}">
        {% csrf_token %}
      </form>
      <form action="{% url 'misago:admin:themes:move-css-down' pk=theme.pk css_pk=item.pk %}" method="post" id="move-down-{{ item.pk }}">
        {% csrf_token %}
      </form>
    {% endfor %}
  {% endwith %}
</div><!-- /.table-panel -->