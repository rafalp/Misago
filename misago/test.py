from django.contrib.messages.api import get_messages
from django.contrib.messages.constants import ERROR, INFO, SUCCESS
from django.test import Client


def assert_contains(response, string, status_code=200):
    assert response.status_code == status_code
    fail_message = f'"{string}" not found in response.content'
    assert string in response.content.decode("utf-8"), fail_message


def assert_not_contains(response, string, status_code=200):
    assert response.status_code == status_code
    fail_message = f'"{string}" was unexpectedly found in response.content'
    assert string not in response.content.decode("utf-8"), fail_message


def assert_has_error_message(response):
    messages = get_messages(response.wsgi_request)
    levels = [i.level for i in messages]

    assert levels, "No messages were set during the request"
    assert ERROR in levels, "No error messages were set during the request"


def assert_has_info_message(response):
    messages = get_messages(response.wsgi_request)
    levels = [i.level for i in messages]

    assert levels, "No messages were set during the request"
    assert INFO in levels, "No info messages were set during the request"


def assert_has_success_message(response):
    messages = get_messages(response.wsgi_request)
    levels = [i.level for i in messages]

    assert levels, "No messages were set during the request"
    assert SUCCESS in levels, "No success messages were set during the request"


def assert_has_message(response, message, level=None):
    messages = get_messages(response.wsgi_request)
    found = False
    for msg in messages:
        if message in str(msg):
            if level and level != msg.level:
                error = (
                    'Message containing "%s" was set '
                    "but didn't have level %s (it had %s)"
                )
                raise AssertionError(error % (message, level, message.level))
            found = True

    if not found:
        raise AssertionError(
            'Message containing "%s" was not set during the request' % message
        )


class MisagoClient(Client):
    def post(self, *args, **kwargs):
        if "json" in kwargs:
            return super().post(
                *args,
                data=kwargs.pop("json"),
                content_type="application/json",
                **kwargs,
            )

        return super().post(*args, **kwargs)

    def put(self, *args, **kwargs):
        if "json" in kwargs:
            return super().put(
                *args,
                data=kwargs.pop("json"),
                content_type="application/json",
                **kwargs,
            )

        return super().put(*args, **kwargs)
