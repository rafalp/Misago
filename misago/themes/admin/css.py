import json
import re

from django.core.files.base import ContentFile

from ...core.utils import get_file_hash


def create_css(theme, css):
    order = None
    if css_exists(theme, css):
        order = get_css_order(theme, css)
        delete_css(theme, css)
    save_css(theme, css, order)


def css_exists(theme, css):
    return theme.css.filter(name=css.name).exists()


def get_css_order(theme, css):
    return theme.css.get(name=css.name).order


def delete_css(theme, css):
    theme.css.get(name=css.name).delete()


def save_css(theme, css, order=None):
    if order is None:
        order = get_next_css_order(theme)

    theme.css.create(
        name=css.name,
        source_file=css,
        source_hash=get_file_hash(css),
        source_needs_building=css_needs_rebuilding(css),
        size=css.size,
        order=order,
    )


def css_needs_rebuilding(css):
    css.seek(0)
    css_source = css.read().decode("utf-8")
    return "url(" in css_source


def get_theme_media_map(theme):
    media_map = {}
    for media in theme.media.all():
        escaped_url = json.dumps(media.file.url)
        media_map[media.name] = escaped_url

        media_filename = str(media.file).split("/")[-1]
        media_map[media_filename] = escaped_url
    return media_map


def rebuild_css(media_map, css):
    if css.build_file:
        css.build_file.delete(save=False)

    css_source = css.source_file.read().decode("utf-8")
    build_source = change_css_source(media_map, css_source).encode()

    build_file_name = css.name
    if css.source_hash in build_file_name:
        build_file_name = build_file_name.replace(".%s" % css.source_hash, "")
    build_file = ContentFile(build_source, build_file_name)

    css.build_file = build_file
    css.build_hash = get_file_hash(build_file)
    css.size = len(build_source)
    css.save()


CSS_URL_REGEX = re.compile(r"url\((.+)\)")


def change_css_source(media_map, css_source):
    url_replacer = get_url_replacer(media_map)
    return CSS_URL_REGEX.sub(url_replacer, css_source).strip()


def get_url_replacer(media_map):
    def replacer(matchobj):
        url = matchobj.group(1).strip("\"'").strip()
        if is_url_absolute(url):
            return matchobj.group(0)

        media_name = url.split("/")[-1]
        if media_name in media_map:
            return "url(%s)" % media_map[media_name]

        return matchobj.group(0)

    return replacer


def is_url_absolute(url):
    if url.startswith("//") or url.startswith("://"):
        return True

    if url.lower().startswith("https://"):
        return True

    if url.lower().startswith("http://"):
        return True

    return False


def get_next_css_order(theme):
    last_css = theme.css.order_by("order").last()
    if last_css:
        return last_css.order + 1
    return 0


def move_css_up(theme, css):
    previous_css = theme.css.filter(order__lt=css.order).order_by("-order").first()
    if not previous_css:
        return False

    css.order, previous_css.order = previous_css.order, css.order
    css.save(update_fields=["order"])
    previous_css.save(update_fields=["order"])

    return True


def move_css_down(theme, css):
    next_css = theme.css.filter(order__gt=css.order).order_by("order").first()
    if not next_css:
        return False

    css.order, next_css.order = next_css.order, css.order
    css.save(update_fields=["order"])
    next_css.save(update_fields=["order"])

    return True
