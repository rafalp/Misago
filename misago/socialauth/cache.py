from django.core.cache import cache

from . import SOCIALAUTH_CACHE
from ..cache.versions import invalidate_cache


def get_socialauth_cache(cache_versions):
    key = get_cache_key(cache_versions)
    return cache.get(key)


def set_socialauth_cache(cache_versions, socialauth):
    key = get_cache_key(cache_versions)
    cache.set(key, socialauth)


def get_cache_key(cache_versions):
    return "%s_%s" % (SOCIALAUTH_CACHE, cache_versions[SOCIALAUTH_CACHE])


def clear_socialauth_cache():
    invalidate_cache(SOCIALAUTH_CACHE)
