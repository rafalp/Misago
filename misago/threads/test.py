from datetime import timedelta

from django.contrib.auth import get_user_model
from django.utils import timezone

from ..acl.test import patch_user_acl
from ..categories.models import Category
from ..core.utils import slugify
from ..users.test import create_test_user
from .checksums import update_post_checksum
from .models import Poll, Post, Thread

default_category_acl = {
    "can_see": 1,
    "can_browse": 1,
    "can_see_all_threads": 1,
    "can_see_own_threads": 0,
    "can_hide_threads": 0,
    "can_approve_content": 0,
    "can_edit_posts": 0,
    "can_hide_posts": 0,
    "can_hide_own_posts": 0,
    "can_merge_threads": 0,
    "can_close_threads": 0,
}


def patch_category_acl(acl_patch=None):
    def patch_acl(_, user_acl):
        category = Category.objects.get(slug="first-category")
        category_acl = user_acl["categories"][category.id]
        category_acl.update(default_category_acl)
        if acl_patch:
            category_acl.update(acl_patch)
        cleanup_patched_acl(user_acl, category_acl, category)

    return patch_user_acl(patch_acl)


def patch_other_user_category_acl(acl_patch=None):
    def patch_acl(user, user_acl):
        if user.slug != "other_user":
            return

        category = Category.objects.get(slug="first-category")
        category_acl = user_acl["categories"][category.id]
        category_acl.update(default_category_acl)
        if acl_patch:
            category_acl.update(acl_patch)
        cleanup_patched_acl(user_acl, category_acl, category)

    return patch_user_acl(patch_acl)


def patch_other_category_acl(acl_patch=None):
    def patch_acl(_, user_acl):
        src_category = Category.objects.get(slug="first-category")
        category_acl = user_acl["categories"][src_category.id].copy()

        dst_category = Category.objects.get(slug="other-category")
        user_acl["categories"][dst_category.id] = category_acl

        category_acl.update(default_category_acl)
        if acl_patch:
            category_acl.update(acl_patch)

        cleanup_patched_acl(user_acl, category_acl, dst_category)

    return patch_user_acl(patch_acl)


def patch_private_threads_acl(acl_patch=None):
    def patch_acl(_, user_acl):
        category = Category.objects.private_threads()
        category_acl = user_acl["categories"][category.id]
        category_acl.update(default_category_acl)
        if acl_patch:
            category_acl.update(acl_patch)
        cleanup_patched_acl(user_acl, category_acl, category)

    return patch_user_acl(patch_acl)


def other_user_cant_use_private_threads(user, user_acl):
    if user.slug == "other-user":
        user_acl.update({"can_use_private_threads": False})


def create_category_acl_patch(category_slug, acl_patch):
    def created_category_acl_patch(_, user_acl):
        category = Category.objects.get(slug=category_slug)
        category_acl = user_acl["categories"].get(category.id, {})
        category_acl.update(default_category_acl)
        if acl_patch:
            category_acl.update(acl_patch)
        cleanup_patched_acl(user_acl, category_acl, category)

    return created_category_acl_patch


def cleanup_patched_acl(user_acl, category_acl, category):
    visible_categories = user_acl["visible_categories"]
    browseable_categories = user_acl["browseable_categories"]

    if not category_acl["can_see"] and category.id in visible_categories:
        visible_categories.remove(category.id)

    if not category_acl["can_see"] and category.id in browseable_categories:
        browseable_categories.remove(category.id)

    if not category_acl["can_browse"] and category.id in browseable_categories:
        browseable_categories.remove(category.id)

    if category_acl["can_see"] and category.id not in visible_categories:
        visible_categories.append(category.id)

    if category_acl["can_browse"] and category.id not in browseable_categories:
        browseable_categories.append(category.id)


User = get_user_model()


def post_thread(
    category,
    title="Test thread",
    poster="Tester",
    is_global=False,
    is_pinned=False,
    is_unapproved=False,
    is_hidden=False,
    is_closed=False,
    started_on=None,
):
    started_on = started_on or timezone.now()

    kwargs = {
        "category": category,
        "title": title,
        "slug": slugify(title),
        "started_on": started_on,
        "last_post_on": started_on,
        "is_unapproved": is_unapproved,
        "is_hidden": is_hidden,
        "is_closed": is_closed,
    }

    if is_global:
        kwargs["weight"] = 2
    elif is_pinned:
        kwargs["weight"] = 1

    try:
        kwargs.update(
            {
                "starter": poster,
                "starter_name": poster.username,
                "starter_slug": poster.slug,
                "last_poster": poster,
                "last_poster_name": poster.username,
                "last_poster_slug": poster.slug,
            }
        )
    except AttributeError:
        kwargs.update(
            {
                "starter_name": poster,
                "starter_slug": slugify(poster),
                "last_poster_name": poster,
                "last_poster_slug": slugify(poster),
            }
        )

    thread = Thread.objects.create(**kwargs)
    reply_thread(
        thread,
        poster=poster,
        posted_on=started_on,
        is_hidden=is_hidden,
        is_unapproved=is_unapproved,
    )

    return thread


def reply_thread(
    thread,
    poster="Tester",
    message="I am test message",
    is_unapproved=False,
    is_hidden=False,
    is_event=False,
    is_protected=False,
    has_reports=False,
    has_open_reports=False,
    posted_on=None,
):
    posted_on = posted_on or thread.last_post_on + timedelta(minutes=5)

    kwargs = {
        "category": thread.category,
        "thread": thread,
        "original": message,
        "parsed": message,
        "checksum": "nope",
        "posted_on": posted_on,
        "updated_on": posted_on,
        "is_event": is_event,
        "is_unapproved": is_unapproved,
        "is_hidden": is_hidden,
        "is_protected": is_protected,
        "has_reports": has_reports,
        "has_open_reports": has_open_reports,
    }

    try:
        kwargs.update({"poster": poster, "poster_name": poster.username})
    except AttributeError:
        kwargs.update({"poster_name": poster})

    post = Post.objects.create(**kwargs)

    update_post_checksum(post)
    post.save()

    thread.synchronize()
    thread.save()
    thread.category.synchronize()
    thread.category.save()

    return post


def post_poll(thread, poster):
    poll = Poll.objects.create(
        category=thread.category,
        thread=thread,
        poster=poster,
        poster_name=poster.username,
        poster_slug=poster.slug,
        question="Lorem ipsum dolor met?",
        choices=[
            {"hash": "aaaaaaaaaaaa", "label": "Alpha", "votes": 1},
            {"hash": "bbbbbbbbbbbb", "label": "Beta", "votes": 0},
            {"hash": "gggggggggggg", "label": "Gamma", "votes": 2},
            {"hash": "dddddddddddd", "label": "Delta", "votes": 1},
        ],
        allowed_choices=2,
        votes=4,
    )

    # one user voted for Alpha choice
    try:
        user = User.objects.get(slug="user")
    except User.DoesNotExist:
        user = create_test_user("User", "user@example.com")

    poll.pollvote_set.create(
        category=thread.category,
        thread=thread,
        voter=user,
        voter_name=user.username,
        voter_slug=user.slug,
        choice_hash="aaaaaaaaaaaa",
    )

    # test user voted on third and last choices
    poll.pollvote_set.create(
        category=thread.category,
        thread=thread,
        voter=poster,
        voter_name=poster.username,
        voter_slug=poster.slug,
        choice_hash="gggggggggggg",
    )
    poll.pollvote_set.create(
        category=thread.category,
        thread=thread,
        voter=poster,
        voter_name=poster.username,
        voter_slug=poster.slug,
        choice_hash="dddddddddddd",
    )

    # somebody else voted on third option before being deleted
    poll.pollvote_set.create(
        category=thread.category,
        thread=thread,
        voter_name="deleted",
        voter_slug="deleted",
        choice_hash="gggggggggggg",
    )

    return poll


def like_post(post, liker=None, username=None):
    if not post.last_likes:
        post.last_likes = []

    if liker:
        like = post.postlike_set.create(
            category=post.category,
            thread=post.thread,
            liker=liker,
            liker_name=liker.username,
            liker_slug=liker.slug,
        )

        post.last_likes = [
            {"id": liker.id, "username": liker.username}
        ] + post.last_likes
    else:
        like = post.postlike_set.create(
            category=post.category,
            thread=post.thread,
            liker_name=username,
            liker_slug=slugify(username),
        )

        post.last_likes = [{"id": None, "username": username}] + post.last_likes

    post.likes += 1
    post.save()

    return like
