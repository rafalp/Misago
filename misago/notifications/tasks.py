from logging import getLogger

from celery import shared_task
from django.contrib.auth import get_user_model

from ..cache.versions import get_cache_versions
from ..conf.dynamicsettings import DynamicSettings
from ..core.pgutils import chunk_queryset
from ..users.bans import get_user_ban
from ..threads.models import Post, Thread
from .models import WatchedThread
from .threads import (
    notify_participant_on_new_private_thread,
    notify_watcher_on_new_thread_reply,
)

NOTIFY_CHUNK_SIZE = 20

User = get_user_model()
logger = getLogger("misago.notifications")


@shared_task(serializer="json")
def notify_on_new_thread_reply(
    reply_id: int,
):
    post = Post.objects.select_related("poster", "thread", "category").get(id=reply_id)
    post.thread.category = post.category

    if not WatchedThread.objects.filter(thread=post.thread).exists():
        return  # Nobody is watching this thread, stop

    cache_versions = get_cache_versions()
    settings = DynamicSettings(cache_versions)

    queryset = WatchedThread.objects.filter(thread=post.thread).select_related("user")

    for watched_thread in chunk_queryset(queryset, NOTIFY_CHUNK_SIZE):
        if (
            watched_thread.user == post.poster
            or not watched_thread.user.is_active
            or get_user_ban(watched_thread.user, cache_versions)
        ):
            continue  # Skip poster and banned or inactive watchers

        try:
            notify_watcher_on_new_thread_reply(
                watched_thread, post, cache_versions, settings
            )
        except Exception:
            logger.exception("Unexpected error in 'notify_watcher_on_new_thread_reply'")


@shared_task(serializer="json")
def notify_on_new_private_thread(
    actor_id: int,
    thread_id: int,
    participants: list[int],
):
    actor = User.objects.filter(id=actor_id).first()
    thread = (
        Thread.objects.filter(id=thread_id)
        .select_related("category", "first_post")
        .first()
    )

    if not actor or not thread:
        return  # Actor or thread could not be found but we need both

    cache_versions = get_cache_versions()
    settings = DynamicSettings(cache_versions)

    queryset = User.objects.filter(id__in=participants)

    for participant in chunk_queryset(queryset, NOTIFY_CHUNK_SIZE):
        if not participant.is_active or get_user_ban(participant, cache_versions):
            continue  # Skip inactive or banned participants

        try:
            notify_participant_on_new_private_thread(
                participant, actor, thread, cache_versions, settings
            )
        except Exception:
            logger.exception(
                "Unexpected error in 'notify_participant_on_new_private_thread'"
            )


@shared_task(serializer="json")
def delete_duplicate_watched_threads(thread_id: int):
    # Merge send emails preference for watched threads
    email_notifications_users_ids = WatchedThread.objects.filter(
        thread_id=thread_id,
        send_emails=True,
    ).values("user_id")

    WatchedThread.objects.filter(
        thread_id=thread_id,
        send_emails=False,
        user_id__in=email_notifications_users_ids,
    ).update(send_emails=True)

    # Delete duplicate watched threads
    kept_watched_threads_ids = (
        WatchedThread.objects.filter(
            thread_id=thread_id,
        )
        .order_by("user_id", "-read_at")
        .distinct("user_id")
        .values("id")
    )

    WatchedThread.objects.filter(thread_id=thread_id).exclude(
        id__in=kept_watched_threads_ids
    ).delete()
