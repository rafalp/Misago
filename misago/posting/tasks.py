from logging import getLogger

from celery import shared_task

from ..threads.models import Post
from . import upgradepost

logger = getLogger("misago.posting")


@shared_task(
    name="posting.upgrade-post-content",
    autoretry_for=(Post.DoesNotExist,),
    default_retry_delay=10,
    time_limit=20,
    serializer="json",
)
def upgrade_post_content(post_id: int, checksum: str):
    post = Post.objects.get(id=post_id)
    if post.sha256_checksum != checksum:
        return

    try:
        upgradepost.upgrade_post_content(post)
    except Exception:
        logger.exception("Unexpected error in 'upgrade_post_content'")
