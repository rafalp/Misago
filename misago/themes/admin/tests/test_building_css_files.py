import pytest

from ....cache.test import assert_invalidates_cache
from ... import THEME_CACHE
from ..css import change_css_source, get_theme_media_map, rebuild_css
from ..tasks import build_single_theme_css, build_theme_css


@pytest.fixture
def assert_snapshot_match(snapshot, theme):
    def _assert_snapshot_match(result):
        result = result.replace(theme.dirname, "themedir")
        assert snapshot == result

    return _assert_snapshot_match


@pytest.fixture
def media_map(theme, image):
    return get_theme_media_map(theme)


def test_tasks_builds_single_css_file(theme, image, css_needing_build):
    build_single_theme_css(css_needing_build.pk)
    css_needing_build.refresh_from_db()
    assert css_needing_build.build_file


def test_tasks_skips_single_css_file_that_doesnt_require_build(theme, css):
    build_single_theme_css(css.pk)
    css.refresh_from_db()
    assert not css.build_file


def test_tasks_handles_nonexisting_css_file(db):
    build_single_theme_css(1)


def test_tasks_builds_theme_css_files_that_require_it(theme, image, css_needing_build):
    build_theme_css(theme.pk)
    css_needing_build.refresh_from_db()
    assert css_needing_build.build_file


def test_tasks_skips_theme_css_files_that_dont_require_build(theme, css):
    build_theme_css(theme.pk)
    css.refresh_from_db()
    assert not css.build_file


def test_tasks_handles_nonexisting_theme(nonexisting_theme):
    build_theme_css(nonexisting_theme.pk)


def test_media_map_for_theme_without_any_media_files_returns_empty_dict(theme):
    assert get_theme_media_map(theme) == {}


def test_media_map_for_theme_with_media_files_returns_dict_with_data(
    theme, image, media
):
    assert get_theme_media_map(theme)


def test_css_file_is_build(media_map, css_needing_build):
    rebuild_css(media_map, css_needing_build)
    css_needing_build.refresh_from_db()
    assert css_needing_build.build_file


def test_build_css_file_is_hashed(media_map, css_needing_build):
    rebuild_css(media_map, css_needing_build)
    css_needing_build.refresh_from_db()
    assert css_needing_build.build_hash


def test_build_css_file_includes_hash_in_filename(media_map, css_needing_build):
    rebuild_css(media_map, css_needing_build)
    css_needing_build.refresh_from_db()
    assert css_needing_build.build_hash in str(css_needing_build.build_file)


def test_build_css_file_has_size_set(media_map, css_needing_build):
    rebuild_css(media_map, css_needing_build)
    css_needing_build.refresh_from_db()
    assert css_needing_build.size


def test_simple_url_to_file_is_replaced_with_valid_url(
    assert_snapshot_match, media_map, image
):
    css = ".page-header { background-image: url(%s); }" % image.name
    result = change_css_source(media_map, css)
    assert_snapshot_match(result)


def test_relative_url_to_file_is_replaced_with_valid_url(
    assert_snapshot_match, media_map, image
):
    css = ".page-header { background-image: url(./%s); }" % image.name
    result = change_css_source(media_map, css)
    assert_snapshot_match(result)


def test_url_to_file_from_create_react_app_is_replaced_with_valid_url(
    assert_snapshot_match, media_map, image
):
    hashed_name = str(image.file).split("/")[-1]
    css = ".page-header { background-image: url(/static/media/%s); }" % hashed_name
    result = change_css_source(media_map, css)
    assert_snapshot_match(result)


def test_quoted_url_to_file_is_replaced_with_valid_url(
    assert_snapshot_match, media_map, image
):
    css = '.page-header { background-image: url("%s"); }' % image.name
    result = change_css_source(media_map, css)
    assert_snapshot_match(result)


def test_single_quoted_url_to_file_is_replaced_with_valid_url(
    assert_snapshot_match, media_map, image
):
    css = ".page-header { background-image: url('%s'); }" % image.name
    result = change_css_source(media_map, css)
    assert_snapshot_match(result)


def test_absolute_https_url_to_file_is_not_replaced(media_map):
    css = ".page-header { background-image: url(https://cdn.example.com/bg.png); }"
    result = change_css_source(media_map, css)
    assert result == css


def test_absolute_http_url_to_file_is_not_replaced(media_map):
    css = ".page-header { background-image: url(http://cdn.example.com/bg.png); }"
    result = change_css_source(media_map, css)
    assert result == css


def test_absolute_protocol_relative_url_to_file_is_not_replaced(media_map):
    css = ".page-header { background-image: url(://cdn.example.com/bg.png); }"
    result = change_css_source(media_map, css)
    assert result == css


def test_css_file_with_multiple_different_urls_is_correctly_replaced(
    assert_snapshot_match, media_map, image
):
    css = (
        ".page-header { background-image: url(http://cdn.example.com/bg.png); }"
        '\n.container { background-image: url("%s"); }'
        '\n.alert { background-image: url("%s"); }'
    ) % (image.name, str(image.file).strip("/")[-1])

    result = change_css_source(media_map, css)
    assert_snapshot_match(result)


def test_building_single_theme_css_invalidates_theme_cache(
    theme, image, css_needing_build
):
    with assert_invalidates_cache(THEME_CACHE):
        build_single_theme_css(css_needing_build.pk)


def test_building_theme_css_invalidates_theme_cache(theme):
    with assert_invalidates_cache(THEME_CACHE):
        build_theme_css(theme.pk)
