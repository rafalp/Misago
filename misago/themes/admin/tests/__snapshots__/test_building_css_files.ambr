# serializer version: 1
# name: test_css_file_with_multiple_different_urls_is_correctly_replaced
  '''
  .page-header { background-image: url(http://cdn.example.com/bg.png); }
  .container { background-image: url("/media/themes/themedir/media/test.357c2ee3.png"); }
  .alert { background-image: url("g"); }
  '''
# ---
# name: test_quoted_url_to_file_is_replaced_with_valid_url
  '.page-header { background-image: url("/media/themes/themedir/media/test.357c2ee3.png"); }'
# ---
# name: test_relative_url_to_file_is_replaced_with_valid_url
  '.page-header { background-image: url("/media/themes/themedir/media/test.357c2ee3.png"); }'
# ---
# name: test_simple_url_to_file_is_replaced_with_valid_url
  '.page-header { background-image: url("/media/themes/themedir/media/test.357c2ee3.png"); }'
# ---
# name: test_single_quoted_url_to_file_is_replaced_with_valid_url
  '.page-header { background-image: url("/media/themes/themedir/media/test.357c2ee3.png"); }'
# ---
# name: test_url_to_file_from_create_react_app_is_replaced_with_valid_url
  '.page-header { background-image: url("/media/themes/themedir/media/test.357c2ee3.png"); }'
# ---
